### PR TITLE
[JIT] canonicalize aten::rsub

### PR DESCRIPTION
**Pull request by mcr229**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65014

Differential Revision: [D30941389](https://our.internmc.facebook.com/intern/diff/D30941389/)

### CODE REVIEW

**Comment by mcr229** [comment]
https://www.internalfb.com/intern/testinfra/testconsole/testrun/4222124719366029/

File: torch/csrc/jit/passes/normalize_ops.cpp [hunk 1]
```diff
@@ -20,6 +20,19 @@ bool normalizeOpAliases(graph_node_list_iterator& iter) {
   return false;
 }
 
+// Normalize rsub such that `rsub(x,y) = sub(x,y)`
+bool normalizeRSub(graph_node_list_iterator& iter) {
+  ArrayRef<Value*> args = iter->inputs();
+  if (iter->kind() == aten::rsub) {
```
Comment (eellison) [inline torch/csrc/jit/passes/normalize_ops.cpp:26]: we all overloads for any particular symbol, so it's possible unlikely that there are other `aten::rsub` operators with different typed inputs for which this substitution isnt valid. just to be safe and future proof we should check the full schema here. you can use the `node->matches("rsub.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor")` api. 

See https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L4880, for this operator what we're doing isnt valid

File: torch/csrc/jit/passes/normalize_ops.cpp [hunk 2]
```diff
@@ -20,6 +20,19 @@ bool normalizeOpAliases(graph_node_list_iterator& iter) {
   return false;
 }
 
+// Normalize rsub such that `rsub(x,y) = sub(x,y)`
+bool normalizeRSub(graph_node_list_iterator& iter) {
+  ArrayRef<Value*> args = iter->inputs();
```
Comment (eellison) [inline torch/csrc/jit/passes/normalize_ops.cpp:25]: nit: move `args` into the if block

File: torch/csrc/jit/passes/normalize_ops.cpp [hunk 3]
```diff
@@ -20,6 +20,19 @@ bool normalizeOpAliases(graph_node_list_iterator& iter) {
...
+  if (iter->kind() == aten::rsub) {
+    iter->replaceWithNewSymbol(aten::sub);
```
Comment (eellison) [inline torch/csrc/jit/passes/normalize_ops.cpp:27]: `replaceWithNewSymbol` returns the updated node, so the replacements we're doing in the following lines will not affect it but the existing node we're destroying

**Review by eellison — COMMENTED** [review]
Looks good!! a few comments from me. test ?

**Comment by mcr229** [comment]
I tried running the test through buck:
 `buck test @mode/dev caffe2/test:jit -- 'test_normalized_rsub \(test_peephole\.TestPeephole\)'`
and it seemed passes, but i don't think it was doing anything, so I will try testing by building and running 
`python3 test/test_jit.py`

**Comment by mcr229** [comment]
Ran tests

`(pytorch) [maxren@devvm3115.atn0 ~/pytorch] python3 test/test_jit.py TestPeephole
CUDA not available, skipping tests
monkeytype is not installed. Skipping tests for Profile-Directed Typing
........s......................
----------------------------------------------------------------------
Ran 31 tests in 0.393s

OK (skipped=1)
(pytorch) [maxren@devvm3115.atn0 ~/pytorch] python3 test/test_jit.py TestPeephole.test_normalized_rsub
CUDA not available, skipping tests
monkeytype is not installed. Skipping tests for Profile-Directed Typing
.
----------------------------------------------------------------------
Ran 1 test in 0.015s

OK
`

File: test/jit/test_peephole.py [hunk 4]
```diff
@@ -219,6 +219,16 @@ def forward(self, x):
             self.run_pass("peephole", conv_dim.graph)
             FileCheck().check("conv").check("dim").run(conv_dim.graph)
 
+    def test_normalized_rsub(self):
+        def convertible_rsub(x: int, y: int):
```
Comment (eellison) [inline test/jit/test_peephole.py:223]: i think we should be checking tensor inputs here

**Review by eellison — COMMENTED** [review]
Nice! just one small comment

File: torch/csrc/jit/passes/normalize_ops.cpp [hunk 5]
```diff
@@ -107,6 +126,7 @@ const std::unordered_map<Symbol, Symbol>& getOperatorAliasMap() {
       {aten::true_divide_, aten::div_},
       {aten::concat, aten::cat},
       {aten::row_stack, aten::vstack},
+      {aten::rsub, aten::sub},
```
Comment (mcr229) [inline torch/csrc/jit/passes/normalize_ops.cpp:129]: the fix for this test
image
seemed to be that rsub could not be found within this map, so adding it to the map and making the appropriate changes seemed to make the test pass

File: test/test_ops.py [hunk 6]
```diff
@@ -775,7 +775,7 @@ def get_sample():
                     #   so running it on all dtypes is would be excessive
                     if dtype == torch.float32:
                         # TODO: no reason why we cant run this with tracing graph
-                        if support_script:
+                        if support_script and op.name != "rsub":
```
Comment (eellison) [inline test/test_ops.py:778]: @alanwaketan this is an example of an annoying test quirk that your PR will fix

**Review by eellison — APPROVED** [review]
Awesome!! 🚢 🚢 🚢

**Pull request merged**